### PR TITLE
[Serve][Doc] Autoscaling

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

- new section of doc for autoscaling (introduction of serve autoscaling and config parameter)
- Remove the version requirement note inside the doc

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
PR description please

File: doc/source/serve/autoscaling.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+(serve-autoscaling)=
+
+# Serve Autoscaling
+
+This section should help you:
+
+- Understand how Ray Serve autoscaling works
+- Understand config parameter usage
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:8]:
```suggestion
- Learn how to configure the parameters for your workload
```

File: doc/source/serve/autoscaling.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+## Autoscaling architecture
+Ray Serve autoscaling is to increase and decrease the number of replicas based on the load, and ray serve provide a couple of config parameters to meet the different workload use cases.
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:12]:
```suggestion
Ray Serve autoscaling feature automatically increases or decreases the number of replicas based on the load. 
```

File: doc/source/serve/autoscaling.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+![pic](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling.svg)
+
+- Deployment handle(Client) and worker replica periodically push the stats to the autoscaler.
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:16]: ServeHandle (link) and deployment replicas (link to managing deployments num_replicas for now, architecture later) periodically push their queuing metrics to the Serve autoscaler.

File: doc/source/serve/autoscaling.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+- Autoscaler requires serve handle queue metrics and replicas requests metrics to make decision whether to scale (up or down) the replicas.
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:17]: Autoscaler uses the ... to make decisions

File: doc/source/serve/autoscaling.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+- Deployment Handle(Client) keeps polling the replica stats from controller to get the updated replicas information. Serve Handle(Client) will send requests directly to the replica based on the replicas information (Round Robin).
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:18]: ServeHandle (link) continues to poll the updated group of replicas from the controller. Upon discovery of the new replicas, it will send any buffered or new requests to the replica until `max_concurrent_queries` is reached (LINK).

File: doc/source/serve/autoscaling.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+:::{note}
+When the controller dies, the client is still able to send requests, but autoscaling will not work.
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:21]: autoscaling will pause to work, when controller recovers, autoscaling will start afresh. All the previous metrics collected will be lost.

File: doc/source/serve/autoscaling.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+:::
+
+## Autoscaling parameters
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:24]:
```suggestion
## Autoscaling parameters
There are several parameters the autoscaling algorithm takes into consideration when deciding the target replicas for your deployment
```

File: doc/source/serve/autoscaling.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+**min_replicas**: The minimal number of replicas for the deployment, the min_replicas will be initial number replicas when the deployment is deployed.
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:25]:
```suggestion
**min_replicas**: The minimal number of replicas for the deployment, the min_replicas will also be the initial number replicas when the deployment is deployed.
```

**Review by simon-mo — COMMENTED** [review]
can you make the autoscaling section in "managing-deployments" shorter and move most of the content to the dedicated page?

**Comment by simon-mo** [comment]
@shrekris-anyscale @architkulkarni @edoakes please review this as well when you get a chance. thanks

**Review by architkulkarni — COMMENTED** [review]
Looks great!  Added some minor grammar/usage suggestions.  

I think this would be even better with some numerical examples for some of the parameters, if they can be explained simply, something like if there are 10 replicas, and config = x, then in the next step we will scale to 12 replicas. (Unless we are trying to hide this from the user in case the underlying implementation changes.)

Also, do we plan to get rid of the existing autoscaling doc in `managing-deployments.md` or leave them both there?  If we leave them both I think there should at least be a link going from the shorter one to the more detailed one.

File: doc/source/serve/autoscaling.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+# Serve Autoscaling
+
+This section should help you:
+
+- Understand how Ray Serve autoscaling works
+- Learn how to configure the parameters for your workload
+
+
+## Autoscaling architecture
+Ray Serve autoscaling feature automatically increases or decreases the number of replicas based on the load.
```
Comment (shrekris-anyscale) [inline doc/source/serve/autoscaling.md:12]:
```suggestion
Ray Serve's autoscaling feature automatically increases or decreases a deployment's number of replicas based on its load. 
```

File: doc/source/serve/autoscaling.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+This section should help you:
+
+- Understand how Ray Serve autoscaling works
+- Learn how to configure the parameters for your workload
+
+
+## Autoscaling architecture
+Ray Serve autoscaling feature automatically increases or decreases the number of replicas based on the load. 
+
+![pic](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling.svg)
```
Comment (shrekris-anyscale) [inline doc/source/serve/autoscaling.md:14]: Nice diagram! This will be very helpful when explaining how Ray Serve's autoscaling works.

Screen Shot 2022-06-13 at 3 55 22 PM

A few suggestions:

1. Can we rename the "node" in the controller to "Ray Head Node"? And similarly, can we name the other nodes to "Ray Worker Node? This is mainly to distinguish between Ray nodes (which are actual machines) and Serve deployment nodes (which are used in deployment graphs).

2. In each Ray node, we have two "worker replica" rectangles. The left rectangle has no outline, but the right rectangle has an outline. Can we remove the outlines on the right rectangles for consistency?

Screen Shot 2022-06-13 at 3 59 33 PM

3. There's a typo in "contoller." It should be "controller."

4. There's a black diagonal "notch" under the controller node. Can we remove it? Can we also move the word "node" in that rectangle, so it doesn't overlap with the arrow?

Screen Shot 2022-06-13 at 4 03 02 PM

5. Can we make the legend left-aligned?

6. The yellow arrow is a bit tough to see. Could we change it (and the corresponding worker replica) to a slightly darker color?

7. To maintain consistency with the rest of the docs, could we change "client (deployment handle)" to "[deployment's `ServeHandle`](https://docs.ray.io/en/releases-1.13.0/serve/http-servehandle.html#servehandle-calling-deployments-from-python)?"

8. Could we add a section to the diagram explaining how HTTP requests fit into this? We could add one more box to the right of the diagram representing an HTTP request, and we could draw an arrow from it to one of the `ServeHandles`.

File: doc/source/serve/autoscaling.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+- Deployment handle(Client) and worker replica periodically push the stats to the autoscaler.
+- Autoscaler requires serve handle queue metrics and replicas requests metrics to make decision whether to scale (up or down) the replicas.
```
Comment (shrekris-anyscale) [inline doc/source/serve/autoscaling.md:17]:
```suggestion
- The autoscaler requires serve handle queue metrics and replicas requests metrics to make decision whether to scale (up or down) the replicas.
```

File: doc/source/serve/autoscaling.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+- Deployment Handle(Client) keeps polling the replica stats from controller to get the updated replicas information. Serve Handle(Client) will send requests directly to the replica based on the replicas information (Round Robin).
```
Comment (shrekris-anyscale) [inline doc/source/serve/autoscaling.md:18]: Let's keep the capitalization and spacing of `ServeHandle` consistent. [The documentation that introduces `ServeHandles`](https://docs.ray.io/en/releases-1.13.0/serve/http-servehandle.html#servehandle-calling-deployments-from-python) uses PascalCase. I suggest we stick with that, but I'm okay with using another format as long as we're consistent.

File: doc/source/serve/autoscaling.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+:::{note}
+When the controller dies, the client is still able to send requests, but autoscaling will pause to work. When the controller recovers, the autoscaling will start working, all previous metrics collected will be lost.
```
Comment (shrekris-anyscale) [inline doc/source/serve/autoscaling.md:21]: In general, we can address the user directly in the docs. For instance, instead of saying "the client is still able to send requests...," we can say "you can still send requests to your deployments..." The [`ServeHandle` documentation](https://docs.ray.io/en/releases-1.13.0/serve/http-servehandle.html#servehandle-calling-deployments-from-python) is a good example for this.

File: doc/source/serve/autoscaling.md [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+:::
+
+## Autoscaling parameters
+There are several parameters the autoscaling algorithm takes into consideration when deciding the target replicas for your deployment
+**min_replicas**: The minimal number of replicas for the deployment, the min_replicas will also be the initial number replicas when the deployment is deployed.
+:::{note}
+Ray Serve Autoscaling allows the `min_replicas` to be 0 to start your deployment, the scale up will be started when you start sending traffic. There will be cold start time during the period, ray serve handle will wait (block) for available replicas to assign the request.
+:::
+**max_replicas**: Max replicas is the maximum number of replicas for the deployment. Ray Serve Autoscaling will rely on the Ray Autoscaler to scale up more nodes when the resource is not enough to bring up more replicas. 
+**target_num_ongoing_requests_per_replica**: The config is to maintain how many ongoing requests concurrently running per replica at most. If the number is lower, the scale up will be done more aggressively.
```
Comment (shrekris-anyscale) [inline doc/source/serve/autoscaling.md:31]: Does `target_num_ongoing_requests_per_replica` set a hard limit on the number of requests a deployment can handle at a certain number of replicas, until a scale-up happens? The phrasing of "at most" makes it seem like that's the case.

**Review by shrekris-anyscale — CHANGES_REQUESTED** [review]
Nice work! These doc updates will be very helpful for new users onboarding to Serve Autoscaling.

File: doc/source/_toc.yml [hunk 15]
```diff
@@ -163,6 +163,7 @@ parts:
               - file: serve/deploying-serve
               - file: serve/monitoring
               - file: serve/performance
+              - file: serve/autoscaling
```
Comment (edoakes) [inline doc/source/_toc.yml:166]: I don't think this should be in the top-level TOC, only the user guides.

File: doc/source/serve/autoscaling.md [hunk 16]
```diff
@@ -0,0 +1,43 @@
+(serve-autoscaling)=
+
+# Serve Autoscaling
+
+This section should help you:
+
+- Understand how Ray Serve autoscaling works
+- Learn how to configure the parameters for your workload
```
Comment (edoakes) [inline doc/source/serve/autoscaling.md:8]:
```suggestion
- Understand how Ray Serve autoscaling works.
- Learn how to configure the parameters for your workload.
```

File: doc/source/serve/autoscaling.md (hunk 3 above)
Comment (edoakes) [reply doc/source/serve/autoscaling.md:16]: We need to make sure our terminology is consistent with the rest of the documentation. Users won't understand what "Deployment handle(Client)" means. Would suggest something like "Each replica with a handle to another deployment periodically pushes stats to the controller."

File: doc/source/_toc.yml (hunk 15 above)
Comment (sihanwang41) [reply doc/source/_toc.yml:166]: All listed here is under `- file: serve/user-guide` if you see more above .

File: doc/source/serve/autoscaling.md (hunk 14 above)
Comment (sihanwang41) [reply doc/source/serve/autoscaling.md:31]: Actually, it is more about the replica view. 
For deployment perspective: 
the number of requests = (running requests + pending requests) * num of replicas.

Comment (architkulkarni) [reply doc/source/serve/autoscaling.md:31]: I second Shreyas's concern, is `target_num_ongoing_requests_per_replica` an upper bound?  From what I remember it's just a target (the true value might be higher or lower), so "at most" might be a little misleading.

**Review by architkulkarni — APPROVED** [review]
Looks good to me!  Just one minor comment about "at most".

File: doc/source/serve/autoscaling.md (hunk 14 above)
Comment (sihanwang41) [reply doc/source/serve/autoscaling.md:31]: I see the point, let me rephrase. Thank you!

File: doc/source/serve/autoscaling.md [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+:::
+
+## Autoscaling parameters
+There are several parameters the autoscaling algorithm takes into consideration when deciding the target replicas for your deployment
+**min_replicas**: The minimal number of replicas for the deployment. ``min_replicas`` will also be the initial number of replicas when the deployment is deployed.
+:::{note}
+Ray Serve Autoscaling allows the `min_replicas` to be 0 when starting your deployment; the scale up will be started when you start sending traffic. There will be a cold start time as the Ray ServeHandle waits (blocks) for available replicas to assign the request.
+:::
+**max_replicas**: Max replicas is the maximum number of replicas for the deployment. Ray Serve Autoscaling will rely on the Ray Autoscaler to scale up more nodes when the currently available cluster resources (CPUs, GPUs, etc) are not enough to bring up more replicas. 
+**target_num_ongoing_requests_per_replica**: The config is to maintain how many ongoing requests are expected to run concurrently per replica at most. If the number is lower, the scale up will be done more aggressively.
```
Comment (shrekris-anyscale) [inline doc/source/serve/autoscaling.md:31]: Nice change! Since this value isn't a hard limit, could we mention that at the end of the description? We could add a sentence saying "Note that this value isn't a hard limit. Under x, y, z circumstances, there could be more ongoing requests per replica than this value."

**Review by shrekris-anyscale — APPROVED** [review]
Great work on all the updates! I left one comment about the `target_num_ongoing_requests_per_replica` parameter.

File: doc/source/serve/autoscaling.md [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+- `target_num_ongoing_requests_per_replica` is only a target value used for autoscaling (not a hard limit), the real ongoing requests number can be higher than the config.
+:::
+**downscale_delay_s**: The config is to control how long the cluster needs to wait before scaling down replicas.
+`upscale_delay_s`: The config is to control how long the cluster need to wait before scaling up replicas.
+:::{note}
+`downscale_delay_s` and `upscale_delay_s` are to control the frequency of doing autoscaling work. E.g. if your use case takes a long time to do initialization work, you can increase `downscale_delay_s` to make the down scaling happen slowly.
+:::
+**smoothing_factor**: The multiplicative factor to speedup/slowdown each autoscaling step. E.g. When the use case has high large traffic volume in short period of time, you can increase `smoothing_factor` to scale up the resource quickly.
+
+**metrics_interval_s**: This is control how often each replica sends metrics to the autoscaler. (Normally you don't need to change this config.)
```
Comment (simon-mo) [inline doc/source/serve/autoscaling.md:44]: add default value to all these parameters.

**Review by simon-mo — APPROVED** [review]

**Review by edoakes — APPROVED** [review]
